File: Swift/task1.swift
/* 
Albert Astrom and Sam Polyakov
CS333
Project 3
task1.swift
*
TO RUN: swift task1.swift
*/

// 1. Identifier naming

let aVariable10000 = "valid" // Can include underscores within or at the beginning of the identifier
print(aVariable10000)
let _under_score_variable = "valid" // Can start with an underscore and have one within the identifier.
print(_under_score_variable)

// let 10000variable = "not valid" // Can't start with a number.

let 👤 = "user" // Can use emoji as identifiers
print(👤)
let θ = 90 // and any unicode character
print(θ)

// 2. Variable declarations:

// Variables are mutable types using the 'var' keyword. 
var mutable = "changeable"
print("mutable: \(mutable)")

mutable = "changed"
print("mutable: \(mutable)")

// Constants are immutable types that are declared using the 'let' keyword.
let immutable = "unchangeable"
// immutableVariable = "unchanged" // This is not allowed and would cause an error.

// 3. Identifier scoping:
// The visibility of an identifier depends on where it is declared.

// Function level scope
func scopeTest() {
    let funcVar = "I am visible only within this function."
    print(funcVar)
    // Global level scope
    print(mutable) // Global variables are visible inside functions.
}

// print(localVariable)  // This would be an error, as localVariable is not visible here.

scopeTest()

// Block level scope
if true {
    let blockVar = "visible"
    print(blockVar)
}

// print(blockVar)  // This would be an error, as blockScopedVariable is not visible here.

File: Rust/task1.rs
/* 
Albert Astrom and Sam Polyakov
CS333
Project 3
task1.rs
*
TO COMPILE: rustc task1.rs
TO RUN: ./task1
*/


fn main() {
    // comparing mutable and immutable variables
    let my_variable = 10; 
    let mut my_mutable_variable = 20; 
    
    // identifiers are case sensitive, variables should start with a lowercase but can start with uppercase
    let MyVariable = 30;
    
    // Identifier names can have underscores
    let my_long_variable_name = 40;
    
    // Identifier names cannot start with a number
    // let 1variable = 50; 
    
    {
        let inner_variable = 60; // Only accessible within this block
        println!("Inner variable: {}", inner_variable);
        
        let my_variable = 70; // Shadows my_variable
        println!("my_variable: {}", my_variable);
    }
    
    // Cannot do this as inner_variable is out of the blocks scope
    // println!("Inner variable: {}", inner_variable);
    
    println!("my_variable: {}", my_variable);
    println!("MyVariable: {}", MyVariable);
    println!("my_long_variable_name: {}", my_long_variable_name);
    
    // cannot modify nonmutable variables
    //my_variable += 10;
    println!("my_variable: {}", my_variable);

    println!("my_mutable_variable: {}", my_mutable_variable);
    // CAN modify mutable variables
    my_mutable_variable += 10;
    println!("my_mutable_variable: {}", my_mutable_variable);
}
